flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5581,10 +5628,210 @@ class ArrayExprLowering {
+      auto exv = std::visit(
+          Fortran::common::visitors{
+              [&](const Fortran::evaluate::IndirectSubscriptIntegerExpr &e)
+                  -> mlir::Value {
+                if (e.value().Rank() == 0)
+                  return fir::getBase(asScalarArray(e.value()));
+                dim++;
+                return fir::getBase(genarr(e.value())(iters));
+              },
+              [&](const Fortran::evaluate::Triplet &t) -> mlir::Value {
+                auto impliedIter = iters.iterValue(dim++);
+                // FIXME: initial should be the lbound of this array. Use 1. See
+                // getLBound().
+                auto initial = builder.createIntegerConstant(loc, idxTy, 1);
+                if (auto optLo = t.lower()) {
+                  auto lo = fir::getBase(asScalarArray(*optLo));
+                  initial = builder.createConvert(loc, idxTy, lo);
+                }
+                auto stride = fir::getBase(asScalarArray(t.stride()));
+                auto step = builder.createConvert(loc, idxTy, stride);
+                auto prod =
+                    builder.create<mlir::MulIOp>(loc, impliedIter, step);
+                auto trip = builder.create<mlir::AddIOp>(loc, initial, prod);
+                return trip;
+              }},
+          sub.u);
+      result.push_back(builder.createConvert(loc, idxTy, fir::getBase(exv)));
+    };
+    auto pushAllIters = [&]() {
+      // FIXME: Need to handle user-defined lower bound. Assume it is the
+      // default, 1.
+      auto one = builder.createIntegerConstant(loc, idxTy, 1);
+      assert(dim == 0 && "array with no subscripts found, but "
+                         "implied iterations already used");
+      for (auto v : iters.iterVec()) {
+        auto vi = builder.createConvert(loc, idxTy, v);
+        result.push_back(builder.create<mlir::AddIOp>(loc, vi, one));
+      }
+      dim += iters.iterVec().size();
+    };
+    for (const auto &v : llvm::reverse(revPath)) {
+      std::visit(Fortran::common::visitors{
+                     [&](ImplicitSubscripts) {
+                       pushAllIters();
+                       ty = fir::unwrapSequenceType(ty);
+                     },
+                     [&](EndOfSubscripts) { ty = fir::unwrapSequenceType(ty); },
+                     [&](const Fortran::evaluate::Subscript *x) { addSub(*x); },
+                     [&](const Fortran::evaluate::ArrayRef *x) {
+                       assert(!x->base().IsSymbol());
+                       if (x->subscript().empty()) {
+                         pushAllIters();
+                       } else {
+                         for (const auto &sub : x->subscript())
+                           addSub(sub);
+                       }
+                       ty = fir::unwrapSequenceType(ty);
+                     },
+                     [&](const Fortran::evaluate::Component *x) {
+                       ty = addField(*x);
```

Comment:
If the component is an array, shouldn't the implicit iteration subscript  be pushed here after the field ?

Reply:
Not the way the list is built. `a%b(i,j)` builds as `[ end, sub(j), sub(i), comp(b), sym(a)]`.